packages/tre/src/storage2/blob/store.ts

```diff
@@ -0,0 +1,191 @@
+import assert from "assert";
+import crypto from "crypto";
+import fs from "fs/promises";
+import path from "path";
+import { arrayBuffer } from "stream/consumers";
+import { ReadableStream } from "stream/web";
+import {
+  InclusiveRange,
+  MultipartOptions,
+  MultipartReadableStream,
+  createArrayReadableStream,
+  createFileReadableStream,
+  createFileWritableStream,
+  createMultipartArrayReadableStream,
+  createMultipartFileReadableStream,
+} from "./streams";
+
+const MODE_READ_ONLY = 0o444;
+
+// Serialisable, opaque, unguessable blob identifier
+export type BlobId = string;
+
+export interface BlobStore {
+  // Database for binary large objects. Provides single and multi-ranged
+  // streaming reads and writes.
+  //
+  // Blobs have unguessable identifiers, can be deleted, but are otherwise
+  // immutable. These properties make it possible to perform atomic updates with
+  // the SQLite metadata store. No other operations will be able to interact
+  // with the blob until it's committed to the metadata store, because they
+  // won't be able to guess the ID, and we don't allow listing blobs.
+  //
+  // For example, if we put a blob in the store, then fail to insert the blob ID
+  // into the SQLite database for some reason during a transaction (e.g.
+  // `onlyIf` condition failed), no other operations can read that blob because
+  // the ID is lost (we'll just background-delete the blob in this case).
+
+  get(
+    id: BlobId,
+    range?: InclusiveRange
+  ): Promise<ReadableStream<Uint8Array> | null>;
+  get(
+    id: BlobId,
+    ranges: InclusiveRange[],
+    opts: MultipartOptions
+  ): Promise<MultipartReadableStream | null>;
+
+  put(stream: ReadableStream<Uint8Array>): Promise<BlobId>;
+
+  delete(id: BlobId): Promise<void>;
+}
+
+function generateBlobId(): BlobId {
+  const idBuffer = Buffer.alloc(32 + 8);
+  crypto.randomFillSync(idBuffer, 0, 32);
+  idBuffer.writeBigUint64BE(process.hrtime.bigint(), 32);
+  return idBuffer.toString("hex");
+}
+
+export class MemoryBlobStore implements BlobStore {
+  readonly #blobs = new Map<string, Uint8Array>();
+
+  get(
+    id: BlobId,
+    range?: InclusiveRange
+  ): Promise<ReadableStream<Uint8Array> | null>;
+  get(
+    id: BlobId,
+    ranges: InclusiveRange[],
+    opts: MultipartOptions
+  ): Promise<MultipartReadableStream | null>;
+  async get(
+    id: BlobId,
+    ranges?: InclusiveRange | InclusiveRange[],
+    opts?: MultipartOptions
+  ): Promise<ReadableStream<Uint8Array> | MultipartReadableStream | null> {
+    const blob = this.#blobs.get(id);
+    if (blob === undefined) return null;
+    if (Array.isArray(ranges)) {
+      assert(opts !== undefined);
```

Comment:
Can this be encoded in the types somehow? Currently they seem to allow `ranges: InclusiveRange, opts: MultipartOptions` and `ranges: InclusiveRange[], opts: undefined`, both of which should be disallowed, I think

Reply:
Ah, I entirely misread this!